### PR TITLE
Implement periodic exporting metric reader

### LINKED ISSUE #1285: Metrics SDK: Implement Periodic exporting MetricReader
As specified in specs: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#periodic-exporting-metricreader

- This is an implementation of the MetricReader which collects metrics based on a user-configurable time interval, and passes it to the exporter
- Configurable parameters:
   -  exportIntervalMillis
   -  exportTimeoutMillis

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1285

## Changes

Specs: https://github.com/open-telemetry/opentelemetry-specification/blob/99d14047268ff670ea2f3cacf597fa258bc4c11c/specification/metrics/sdk.md#periodic-exporting-metricreader

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+class MetricExporter;
+/**
+ * Struct to hold batch PeriodicExortingMetricReader options.
+ */
+
+struct PeriodicExportingMetricReaderOptions
+{
+
+  /* The time interval between two consecutive exports. */
+  std::chrono::milliseconds schedule_delay_millis = std::chrono::milliseconds(60000);
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h:30]: The name `schedule_delay_millis` might be misleading.

The suggested name is `export_interval_millis` https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#periodic-exporting-metricreader

Comment (reyang) [reply sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h:30]: For example, if export 1 runs during [T1, T2] and export 2 runs during [T3, T4], `export_interval_millis = T3 - T1`, while `schedule_delay_millis = T3 - T2`.

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h]: Thanks, changed the variable name to `export_interval_millis`, and also logic to ensure there is exactly `schedule_delay_millis` ms difference between two exports ( by adjusting the time taken for collect and export operation ).

File: sdk/src/metrics/metric_reader.cc [hunk 2]
```diff
@@ -46,18 +47,22 @@ bool MetricReader::Collect(nostd::function_ref<bool(MetricData)> callback) noexc
...
+    const std::lock_guard<opentelemetry::common::SpinLockMutex> locked(lock_);
+    shutdown_ = true;
+  }
+
   if (!OnShutDown(timeout))
   {
     status = false;
     OTEL_INTERNAL_LOG_WARN("MetricReader::OnShutDown Shutdown failed. Will not be tried again!");
   }
   const std::lock_guard<opentelemetry::common::SpinLockMutex> locked(lock_);
```
Comment (esigo) [inline sdk/src/metrics/metric_reader.cc:65]: `status` is local to the function.

File: sdk/include/opentelemetry/sdk/metrics/metric_exporter.h [hunk 3]
```diff
@@ -32,9 +32,7 @@ class MetricExporter
    * concurrently for the same exporter instance.
    * @param spans a span of unique pointers to metrics data
```
Comment (esigo) [inline sdk/include/opentelemetry/sdk/metrics/metric_exporter.h:33]: comment needs to be updated.

File: sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  std::chrono::milliseconds export_interval_millis_;
+  std::chrono::milliseconds export_timeout_millis_;
+
+  void DoBackgroundWork();
+
+  /* The background worker thread */
+  std::thread worker_thread_;
+
+  /* Synchronization primitives */
+  std::condition_variable cv_, force_flush_cv_;
```
Comment (esigo) [inline sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h:66]: I can't remember if I've seen this style in the repo.
If new, I'd vote for not using it.

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h]: It's taken from here :-) https://github.com/open-telemetry/opentelemetry-cpp/blob/a7e814a632fdaee0c76246358ffad811ac06a7e9/sdk/include/opentelemetry/sdk/trace/batch_span_processor.h#L141 
But still agree it's a bad style. Also realized we don't need cv/mutex for flush and shutdown, as it's handled by [MetricReader](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/sdk/src/metrics/metric_reader.cc). Have removed them now.

File: sdk/include/opentelemetry/sdk/metrics/metric_exporter.h (hunk 3 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/metric_exporter.h]: Thanks fixed.

File: sdk/src/metrics/metric_reader.cc (hunk 2 above)
Comment (lalitb) [reply sdk/src/metrics/metric_reader.cc:65]: Yes, it is meant to be local. Do you see any issue here?

File: sdk/test/metrics/periodic_exporting_metric_reader_test.cc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  bool Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override
+  {
+    return true;
+  }
+
+  size_t GetDataCount() { return records.size(); }
+
+private:
+  std::vector<MetricData> records;
```
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/test/metrics/periodic_exporting_metric_reader_test.cc]: Thanks fixed.

File: sdk/src/metrics/metric_reader.cc (hunk 2 above)
Comment (esigo) [reply sdk/src/metrics/metric_reader.cc:65]: Sorry I meant we don't need the lock in line 65.

**Review by esigo — APPROVED** [review]
LGTM
Thanks for the PR :)

File: sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+namespace metrics
+{
+
+class MetricExporter;
+/**
+ * Struct to hold PeriodicExortingMetricReader options.
+ */
+
+constexpr std::chrono::milliseconds kExportIntervalMillis = std::chrono::milliseconds(60000);
+;
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h:27]:
```suggestion
```

File: sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+{
+namespace metrics
+{
+
+class MetricExporter;
+/**
+ * Struct to hold PeriodicExortingMetricReader options.
+ */
+
+constexpr std::chrono::milliseconds kExportIntervalMillis = std::chrono::milliseconds(60000);
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h:26]: Are the default constant coming from the spec?

**Review by ThomsonTan — APPROVED** [review]

File: sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h (hunk 7 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/export/periodic_exporting_metric_reader.h:26]: yes, it is coming from here: https://github.com/open-telemetry/opentelemetry-specification/blob/99d14047268ff670ea2f3cacf597fa258bc4c11c/specification/metrics/sdk.md#periodic-exporting-metricreader

**Pull request merged**